src/dispatching/middleware.rs

```diff
@@ -0,0 +1,25 @@
+use std::{future::Future, pin::Pin};
+
+/// An asynchronous middleware.
+///
+/// See [the module-level documentation for the design
+/// overview](crate::dispatching).
+pub trait Middleware<T> {
+    #[must_use]
+    fn handle<'a>(&'a self, val: T) -> Pin<Box<dyn Future<Output = T> + 'a>>
+    where
+        T: 'a;
+}
+
+impl<T, F, Fut> Middleware<T> for F
+where
+    F: Fn(T) -> Fut,
+    Fut: Future<Output = T>,
+{
+    fn handle<'a>(&'a self, val: T) -> Pin<Box<dyn Future<Output = T> + 'a>>
```

Comment:
Maybe middleware will return `Option<T>`? In this case:
1. If middleware return Some(T), pass it on.
2. If middleware return None, do nothing with it library.
First, i want use it in future, to create CommandMiddleware, which handle all commands. Second, this allow to create middlewares like CheckBotBanned, which find user in special blacklist and if he is in this list, return None.

Reply:
Ok, I'll rework middlewares.